rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+
+#### Functional Index Metadata
+
+For each functional index, store a separate metadata with index details. This should be efficiently loaded. One option
+is to store the below metadata in `hoodie.properties`. This way all index metadata can be loaded with the table config.
+But, it would be better to not overload the table config. Let `hoodie.properties` still hold the list of indexes (MT
+partitions) available, and we propose to create separate `.hudi_index` directory under `basepath/.hoodie`. This has the
+benefit of separation of concern and can be cheaply loaded with the metaclient (one extra file i/o only if there is a
+functional index in the table config). It can store the following metadata for each index.
+
+* Name of the index
+* Expression (e.g., `MONTH(timestamp)`)
+* Data column(s) it's derived from
+* Any other configuration or properties
+
+#### APIs
+
+A new interface, `HoodieFunctionalIndex`, is introduced to represent a functional index and track the relationship
+between the original data column(s) and the transformation.
+
+```java
+interface HoodieFunctionalIndex<S, T> {
+  String indexName;
+  String expression;
+  List<String> columnsInExpression;
+
+  T apply(S sourceValue);
+}
+```
+
+As mentioned in the goals of this RFC, we are not going to build an expression engine. To begin with, we can simply
+parse `String` for simpler functions and match it to the functional keywords available in spark-sql. We can tackle other
+engines as we go.
+
+#### Index in Metadata Table
+
+For each functional index, Hudi needs to populate the index to facilitate data skipping. We reuse the metadata table and
+each functional index is stored under a new partition named as `func_index_<index_name>`. Each record in index is a
+key-value and payload will depend on `index_type`. For example, if `index_type` is not provided then key is simply the
+values based on function or expression and value is a collection of file paths. If `index_type` is `column_stats` then
+the value will be stats based on value derived from the function and key will be hash of file path. 
+
+For an existing table, one can use the async indexer to build the indexes concurrently with ingestion. When data is
+written:
+
+1. Use the storage strategy (hive-style, random prefix, etc.) to determine the physical location to write the data (
+   which we will do anyway for commit metadata).
+2. Update the index i.e. update MT for each index.
+    1. Synchronous update: Data table commit is not committed until all indexes are updated.
+    2. Async update: Use the [async indexer](https://hudi.apache.org/docs/metadata_indexing) to load the records
+       incrementally and update the index.
+
+#### Query Planning & Execution
+
+On receiving a query:
+
+1. Parse the query and index metadata (in `.hudi_index`) to determine the relevant functional index is available or not.
+2. Lookup (point/range) index to determine:
+    * Which files match functional index predicates
```

Comment:
IMO we need to do a second level aggregation for aggregating index types like column statistics, to quickly filter out storage partitions (e.g partitions within Hive style partitioning storage organization), with the same effect of partition pruning.

Reply:
Do we need to do that for all functional index? Whenever user creates a functional index using column_stats then stats will be aggregated at "partition/file" level. If index_type is not provided then the the value derived by the function is the key and collection of file paths is the value. In that case we don't even have to skip based on statistics. Since keys will be sorted we can directly skip based on predicate.

I also looked at [covering indexes](https://www.postgresql.org/docs/current/indexes-index-only-scans.html) and I think we can add that in the current design. Need to add one more keyword in CREATE INDEX syntax.

Reply:
Updated with an example we discussed offline.

Reply:
I added a line there. but just want to confirm that this storage partition level aggregation is in scope of what we are implementing.

Reply:
Yes, it's in the scope of 1.0.